### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by progfolio**
Added a few unit tests using buttercup. They can be run with `make test` from the straight repo directory. They do assume buttercup is installed, so we'd have to arrange for that if we wanted to run these on CI, but that should be easy to do with straight itself.

image

Related: #478 #69 

Let me know what you think about this setup in general. If you like it, I'll try and get us covered on all the unit tests.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -5,6 +5,9 @@ EMACS ?= emacs
 
 SHELL := bash
 
+#Point to path of your local emacs-buttercup install
+BUTTERCUP = ../emacs-buttercup/
```
Comment (raxod502) [inline Makefile:9]:
```suggestion
BUTTERCUP ?= ../emacs-buttercup/
```
this allows setting it on the command line or by environment variable.

File: tests/test-straight.el [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+(provide 'straight-test)
+
+;; Local Variables:
+;; End:
+
+;;; straight-test.el ends here
+
+
+
+;;; test-straight.el ends here
```
Comment (raxod502) [inline tests/test-straight.el:120]: We can probably ditch most of these headers, like in `bootstrap.el` and `install.el`, as the file isn't part of the package. Would be easier than filling them out with information that would probably be mostly redundant.

File: Makefile [hunk 3]
```diff
@@ -48,7 +51,6 @@ checkdoc: ## Check docstring style
 .PHONY: longlines
 longlines: ## Check for long lines
 	@scripts/check-line-length.bash
-
```
Comment (raxod502) [inline Makefile:51]: I think this line was deleted by accident?

File: tests/test-straight.el [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        (expect (straight--alist-set "a" 'b  '(("c" . d)))
+                :to-equal '(("a" . b) ("c" . d))))
+      (it "sets first KEY match cdr"
+        (expect (straight--alist-set "a" 'd  '(("a" . b) ("a" . c)))
+                :to-equal '(("a" . d) ("a" . c)))))
+    (describe "(key val alist symbol)"
+      (it "tests with eq when SYMBOL non-nil"
+        (expect (straight--alist-set "a" 'a  '((a . b) (a . c)) 'symbol)
+                :to-equal '(("a" . a) (a . b) (a . c))))))
+(describe "straight-vc-git--fork-repo"
```
Comment (raxod502) [inline tests/test-straight.el:45]: Should this be nested under `(describe "straight.el" ...)` as well? I'm not actually familiar enough with Buttercup to say.

**Review by raxod502 — COMMENTED** [review]
Yeah, this looks great!

File: tests/test-straight.el (hunk 2 above)
Comment (progfolio) [reply tests/test-straight.el:120]: I usually leave the bare minimum to satisfy checkdoc/package-lint. Up to you, though.

Comment (raxod502) [reply tests/test-straight.el:117]: Personally, I prefer to either write a proper header, or not have one at all. In this case I think having no header is quite reasonable.

Comment (progfolio) [reply tests/test-straight.el]: Removed

**Review by raxod502 — APPROVED** [review]

**Pull request merged**